subsys: bootloader: fix breaking changes introduced by kconfig after DTC

sigvartmh opened the pull request:
Fixes secure bootloader so that it can be compiled with master for nrf52840 and nrf9160 due to the breaking changes when moving Kconfig after DTC.

sigvartmh commented:
@hakonfam @oyvindronningstad @osaether

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:132: ERROR:C99_COMMENTS: do not use C99 // comments
#132: FILE: subsys/bootloader/bootloader.h:1:
+//TODO: Remove when multi-image has been merged

-:184: ERROR:C99_COMMENTS: do not use C99 // comments
#184: FILE: subsys/bootloader/bootloader.h:53:
+#endif //CONFIG_BOARD_NRF52850_PCA10056

-:185: ERROR:C99_COMMENTS: do not use C99 // comments
#185: FILE: subsys/bootloader/bootloader.h:54:
+#endif // BOOTLOADER_H__

-:220: WARNING:FUNCTION_ARGUMENTS: function definition argument 'void' should also have an identifier name
#220: FILE: subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c:29:
+void z_clock_isr(void) ALIAS("dummy_handler");

-:223: WARNING:FUNCTION_ARGUMENTS: function definition argument 'void' should also have an identifier name
#223: FILE: subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c:32:
+void assert_post_action(void) ALIAS("dummy_func");

- total: 3 errors, 2 warnings, 520 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```

sigvartmh commented:
@hakonfam @oyvindronningstad @osaether

hakonfam commented:
@carlescufi  This i a blocker for all other bootloader related PRs, so a rapid response would be greatly appreciated :)

hakonfam commented on subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c:
```diff
@@ -26,8 +26,14 @@ void __secure_fault(void) ALIAS("dummy_handler");
 void __svc(void) ALIAS("dummy_handler");
 void __pendsv(void) ALIAS("dummy_handler");
 void __sys_tick(void) ALIAS("dummy_handler");
+void z_clock_isr(void) ALIAS("dummy_handler");
```
Could you put this in a separate commit?

sigvartmh replied:
Sure

hakonfam commented:
The commit "subsys: bootloader: cmake: add overide to provision script"  should have different tags as it only changes a python file. "scripts: bootloader:" perhaps.

oyvindronningstad commented on scripts/bootloader/provision.py:
```diff
@@ -55,6 +55,9 @@ def parse_args():
         formatter_class=argparse.RawDescriptionHelpFormatter)
 
     parser.add_argument("--generated-conf-file", required=True, help="Generated conf file.")
+    parser.add_argument("--s0-addr", type=int, required=False, help="Set address of s0 explicit")
```
In next three lines:
explicit -> explicitly

sigvartmh replied:
fixed (not pushed)

oyvindronningstad commented on scripts/bootloader/provision.py:
```diff
@@ -55,6 +55,9 @@ def parse_args():
     parser.add_argument("--generated-conf-file", required=True, help="Generated conf file.")
+    parser.add_argument("--s0-addr", type=int, required=False, help="Set address of s0 explicit")
+    parser.add_argument("--s1-addr", type=int, required=False, help="Set address of s1 explicit")
+    parser.add_argument("--provision-addr", type=int, required=False, help="Set address of proision data explicit")
```
proision -> provision

sigvartmh replied:
fixed(not pushed)

oyvindronningstad commented on subsys/bootloader/CMakeLists.txt:
```diff
@@ -177,13 +169,32 @@ add_custom_target(
   )
 
+if(CONFIG_SOC_NRF52840)
+	set(OVERRIDE
```
I suggest
OVERRIDE -> ADDR_OVERRIDE

sigvartmh replied:
Renamed(not pushed)

oyvindronningstad commented on subsys/bootloader/bootloader.h:
```diff
@@ -0,0 +1,72 @@
+/*TODO: Remove when multi-image has been merged*/
```
Why not put the addresses in Kconfig? That would make them available in CMake as well.

sigvartmh replied:
The only reasoning was less diff from now to multi image. Since the variable names will be prefixed with CONFIG_ instead of DT_. 

This is a fixup until multi-image is merged when multi image is merged removing `bootloader.h` would give you compile errors so it's easier to remove where it's used. Instead of modifying the Kconfig file then revert. All files touched but it's open for discussion if you would rather prefer it placed in Kconfig.

oyvindronningstad replied:
Ok, I was wondering if we could reuse the values for the provision override parameters in the cmake file. It's maybe not worth it at this point, at least if it will change again.

sigvartmh replied:
I don't think it's worth it at this point, as we would have to change it again. So it's better to change it when we have the actual build system in place :1st_place_medal:

oyvindronningstad commented on subsys/bootloader/c_runtime_setup/Kconfig:
```diff
@@ -4,17 +4,9 @@
-endchoice
+config SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
+	bool "Zephyr C runtime setup"
+	default y
```
What happens if this is set to 'n'?

sigvartmh replied:
Compile error, I'll just remove the prompt

sigvartmh replied:
Done

oyvindronningstad commented:
The commit "subsys: bootloader: Remove default Kconfig values" could have a more descriptive commit message:
 - Which default Kconfig values?
 - Why can these choices no longer be used?

hakonfam commented on subsys/bootloader/bootloader.h:
```diff
@@ -0,0 +1,72 @@
+#if CONFIG_SOC_NRF52840
+/* partition@0 */
+#define DT_FLASH_AREA_SECURE_BOOT_OFFSET	0
+#define DT_FLASH_AREA_SECURE_BOOT_SIZE	16384
```
Note that all these are replaced when we get multi image support in the build system.

hakonfam replied:
But I agree :)

sigvartmh replied:
It's DTC output but sure I can try to format them back

oyvindronningstad replied:
Oh, I see, you just took the generated file. In that case it's not a big deal, since it is temporary.

sigvartmh replied:
Fixed(not pushed)

sigvartmh commented:
> The commit "subsys: bootloader: Remove default Kconfig values" could have a more descriptive commit message:
> 
>     * Which default Kconfig values?
> 
>     * Why can these choices no longer be used?

I'll add some more text explaining the rational

sigvartmh commented on subsys/bootloader/bootloader.h:
```diff
@@ -0,0 +1,72 @@
+#define DT_FLASH_AREA_APP_SIZE	929792
+
+/* partition@4000 */
+#define DT_FLASH_AREA_S0_OFFSET		16384
```
Type First->Second->Third->ETC... ?

oyvindronningstad replied:
Just the order they appear in flash.
secure boot -> s0 -> s1 -> app -> provision.
Again, if you just took generated output, it's not a big deal.

sigvartmh replied:
fixed(not pushed)

sigvartmh commented on subsys/bootloader/bootloader.h:
```diff
@@ -0,0 +1,72 @@
+#define DT_FLASH_AREA_S0_SIZE		49152
+
+/* partition@FF000 */
+#define DT_FLASH_AREA_PROVISION_OFFSET	1044480
```
You could do `FLASH_SIZE-PROVISION_SIZE`

oyvindronningstad replied:
Yes, that's what I was thinking. That should be sufficient to support all nRF5x devices.

sigvartmh replied:
Yeah it works for the nrf5 devices. Couldn't find a general config variable for it so it will have to be
```
#if CONFIG_SOC_NRF52840  ||\
    CONFIG_SOC_NRF52832  ||\
    CONFIG_SOC_NRF52810  ||\
    CONFIG_NRF51822_QFAA ||\
    CONFIG_NRF51822_QFAB ||\
    CONFIG_NRF51822_QFAC
```

sigvartmh replied:
Done

sigvartmh commented on subsys/bootloader/bootloader.h:
```diff
@@ -0,0 +1,72 @@
+#define DT_FLASH_AREA_S0_OFFSET	32768
+#define DT_FLASH_AREA_S0_SIZE		16384
+
+/* partition@C0000 */
```
DTC output so did not really check if comments are correct

sigvartmh replied:
Done

sigvartmh commented:
This PR has been **updated**. Most notable changes are done in `bootloader.h` other than that the commit history has been rewritten and split up into logically preceding commits.
*  Kconfig symbol for changing linkerscript/startup code is deprecated and removed
* More unused files are removed hence the ~-4000LOC

oyvindronningstad commented on subsys/bootloader/bootloader.c:
```diff
@@ -86,9 +86,6 @@ void uninit_used_peripherals(void)
 
 #ifdef CONFIG_SW_VECTOR_RELAY
-#ifndef CONFIG_SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
-_GENERIC_SECTION(.vt_pointer_section) u32_t _vector_table_pointer;
```
Are you sure this line is not needed?
It's for nRF51 which has no VTOR.

sigvartmh replied:
I could be wrong but I thought it was not needed if `SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR` is standard and that it defines this `extern  _vector_table_pointer` in some other file am I wrong?

oyvindronningstad replied:
Oh, sorry I misread, it's #ifndef not #ifdef.

sigvartmh commented:
@nvlsianpu This is ready to be reviewed

sigvartmh commented on subsys/bootloader/CMakeLists.txt:
```diff
@@ -177,13 +169,32 @@ add_custom_target(
 
+if(CONFIG_SOC_NRF52840)
+	set(ADDR_OVERRIDE
+		--s0-addr 16384
```
Oh must have overlooked it will change it to hex

sigvartmh replied:
Fixed

sigvartmh commented on subsys/bootloader/include/linker.ld:
```diff
@@ -111,6 +132,17 @@ SECTIONS
 	KEEP(*(".dbghdr.*"))
 #endif
+
+#ifdef CONFIG_NXP_IMX_RT_BOOT_HEADER
```
Yes, this file is temporary until multi-image is merged.

SebastianBoe commented on subsys/bootloader/c_runtime_setup/CMakeLists.txt:
```diff
@@ -52,68 +40,8 @@ target_include_directories_ifdef(CONFIG_SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
-  gcc_startup_nrf9160.S
-  )
-endif()
+add_library(kernel_init STATIC ${ZEPHYR_BASE}/kernel/init.c)
```
This isn't a zephyr library, because it should be linked into the bootloader, and not the app, right?

Please add a comment that explains this.

Then I wouldn't be confused again, and it will be more clear that this needs to change after the multi-image migration.

sigvartmh replied:
Yes, you are correct it isn't a zephyr library it was a hackish way of being able to set `_Cstart=UNUSED_Cstart` only for compiling the C runtime/startup code of zephyr then override the _Cstart rutine with our bootloaders `main_bl` call. I'll add a comment explaining this. Do you mind elaborating a bit on what needs to be changed after multi-image ?

sigvartmh replied:
Added comment @SebastianBoe check if the comment gives enough information so that you are not confused by the purpose of this lib.

SebastianBoe replied:
LGTM

SebastianBoe commented on subsys/bootloader/CMakeLists.txt:
```diff
@@ -177,13 +169,32 @@ add_custom_target(
+		--provision-addr 1044480)
+endif()
+
+
```
DT_FLASH_AREA_S0_OFFSET is now available in the DT namespace (you can do ${DT_FLASH_AREA_S0_OFFSET}).

So I think you want to remove the generated_dts_board.conf parser function, and instead determine the possibly overridden value in CMake. instead of in python. Ask me for details. I could be wrong.

sigvartmh replied:
It's a great idea unfortuntaly these symbols does not yet exsists in the DT https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/1d6219ffec845a43f66469354d997fbea2f283a4/boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts#L165

I'll investigate more into your 2nd comment I'll ask you tomorrow :)

SebastianBoe replied:
I can't read DT source code and predict what symbols will be available ;)

The reason I believed they were available is that when I build, I get symbols like these in CMake:

```
DT_FLASH_AREA_MCUBOOT_OFFSET=0
DT_FLASH_AREA_STORAGE_OFFSET=122880

```

But I haven't tried in a build that has the S0 area defined.

sigvartmh replied:
Trying to predict naming and which symbols are available from DT is quite hard and it was a nice guess the only thing missing to get the symbols you are asking for is 

```
storage_partition: partition@(offset) { //@has no effect and is just inferred as a comment)
			label = "s0";
			reg = <(offset) (size)>;
};
```
in the partition section but this will have to be added for every board instead of every SoC I think? Honestly I'm not sure what would be best, this would at least give the smallest diff towards zephyr if we don't want to try to upstream partitioning changes upstream.

A problem with trying to upstream it is that MCUBoot is relaying on the mcuboot symbols as they can't overlay their own DT overlay file outside the zephyr project folder as overlays have to be board specific to be overlayed

sigvartmh replied:
Moved the code out of python and into cmake now so that instead of an override we do an fallback if the values does not exsist in DT

SebastianBoe replied:
Great!
